tests/test_plugin.py

```diff
@@ -154,6 +154,77 @@ def test_required_plugin_failure(tmpdir, docker_tasker, runner_type, required):
         assert workflow.plugin_failed is required
 
 
+@pytest.mark.parametrize('runner_type, plugin_type', [  # noqa
+    (PreBuildPluginsRunner, PreBuildPlugin),
+    (PrePublishPluginsRunner, PrePublishPlugin),
+    (PostBuildPluginsRunner, PostBuildPlugin),
+    (ExitPluginsRunner, ExitPlugin),
+    (BuildStepPluginsRunner, BuildStepPlugin),
+])
+@pytest.mark.parametrize('required', [
+    True,
+    False,
+])
+def test_verify_required_plugins_before_first_run(caplog, tmpdir, docker_tasker, runner_type,
+                                                  plugin_type, required):
+    """
+    test plugin availability checks before running any plugins
+    """
+    class MyPlugin(plugin_type):
+        key = 'MyPlugin'
+
+        def run(self):
+            return DUMMY_BUILD_RESULT
+
+    flexmock(PluginsRunner, load_plugins=lambda x: {
+                                        MyPlugin.key: MyPlugin})
+    workflow = mock_workflow(tmpdir)
+    # we rely on the plugin run order
+    runner = runner_type(docker_tasker, workflow,
+                         [{"name": MyPlugin.key, "required": False},
+                          {"name": "no_such_plugin", "required": required}])
+    expected_log_message = "running plugin '%s'" % MyPlugin.key
+
+    # build step plugins set "required" to False
+    if required and (runner_type != BuildStepPluginsRunner):
+        with pytest.raises(PluginFailedException):
+            runner.run()
+        assert expected_log_message not in [l.getMessage() for l in caplog.records()]
```

Comment:
Don't waste memory by creating lists :-)
`assert all(expected_log_message not in l.getMessage() for l in caplog.records())`